### ISSUE TITLE
About the mechanics of round 2 and blame

### ISSUE THREAD

**Issue by AdamISZ** [supertestnet/coinjoin-workshop#1]
This kind of follows up some offhand discussion on nostr. TLDR I think you need linkable ring sigs and to think through blame in more detail.

So in round 2 I am assuming:

* the message takes the form (ring signature, coinjoin address). ignoring any trivial metadata is that correct?
* the round 2 messages are sent as a "broadcast" from an anonymised source, like Tor

The second point was not discussed I think? But it wouldn't make sense to send (rs, cj_addr) with an identifiable point of origin (unless you make an assumption about trusting the relay with your privacy, a little like a VPN). So I am not exactly sure how that is intended to work, but for sure it can be done with e.g. Tor, modulo connection flakiness (this has been a fairly severe issue requiring a lot of attention, even for *central coordinators*!).

So, anywa, continuing, let's assume the ring signature is *not* linkable as suggested, that is the primitive we're using is SAG but not LSAG which means it is a spontaneous (you can form the signature without the cooperation of the other keyholders) anonymous (i.e. does not reveal which key) group signature. But it is not *linkable* in that if you publish two signatures using this method and using the same key, the two signatures are not identifiably from the same key.

In this case we have scenarios like (here notation 'r' is a ring signature, 'a' is a coinjoin address):
only 3 participants (certain scenarios will be more or less likely at e.g. 100)

Scenario 1:
alice posts (r1, a1)
charlie posts (r2, a2)
bob posts (r3, a3)

This is the happy case. Everyone sees 3 ring sigs and constructs the tx and goes on to round 3 correctly.

Scenario 2:
alice posts (r1, a1)
charlie posts (r2, a2)
alice posts (r3, a3)
bob posts (r4, a4)

This is just an example of malicious activity, but just to mention the obvious to other readers: trying to "steal" like this in coinjoin protocols never works (at least never succeeds in stealing!) ... we only sign transactions that don't steal from us (i.e. the only risk is a coding logic error). But let's think about blame:

So let's ask how the code decides to terminate round 2. Does it just wait for N messages (here 3)? If so, it will terminate on receiving (r3, a3) and move on to round 2, ignoring (r4, a4). Again, since they are anonymous, anyone could send anything at any time, so that certainly seems the most reasonable(?).

If so then in round 3 bob will not sign, of course. Is Bob kicked out for refusing to send a transaction-signing message? 

(btw, I think questions like this can arise because the way you framed Round 3 is a bit off: you start with "suppose a troll did round 1 and round 2 correctly and is now in round 3" - precisely *if* we don't use linkable ring sigs, we can't know for sure that they did round 2 correctly)

Alternatively, perhaps we treat it as being the case that if there is any failure at round 3 we have to go back and "open" the ring signatures from round 2; that would be logical.

But that comes back to what I want to mention below about "claimability" and what you referred to as "revealing the private key for the ring signature":

I think examples like this illustrate two things:

* we have to be precise about the mechanics of "revealing the key for the ring signature", and
* linkable ring signatures are required here

Actually those two points are kind of connected/the same. So, a little ring signature deep lore here: we want "claimability" as a property.

Linkable ring signatures (I think always?) have claimability built in: along with a Schnorr style signature (most but not all ring sig designs use it as a basis), you attach a "key image" (sometimes "linking tag"), it's nothing too sci-fi it's just (private key)*(other basepoint) ('other' because not the usual basepoint 'G'). In effect it's an additional 32 byte tag that will never change even when you change the message you're signing. How that other basepoint is calculated varies per design but it can be e.g. hash-to-curve(list-of-all-keys-in-ring).

The nice thing about this design for our purposes is that all you have to do to reveal that a ring signature came from your pubkey, is to reveal your private key (so it lines up with your description in the README); the everyone else can calculate your-private-key * other-basepoint to get your key image and verify where it was used.

As an alternative non-linkable, consider AOS 2004 (see my [blog](https://reyify.com/blog/ring-signatures) for some background and links on that), where you just form the ring signature as a "coupled" set of Schnorr equations; doing it this way, I'm not sure how easily you can prove it was your specific private key that signed it;  *simply revealing your private key* on its own doesn't do it, I think (in contrast with the previous paragraph, note).

(Another thing to consider: scalability is not a trivial issue here if you want to go to 100 participants and higher. It *probably* won't matter but you at least have to consider it, because the "normal" ring sig designs are linear in the keyset, e.g. 2*32*n for n keys (and there might be some n^2 broadcast going on too). Maybe that isn't an issue at all, but if it is, there are designs that are much more efficient and not *too* exotic.)

Finally, coming back to the trickiness of getting blame right, I think this quote from the original [Coinshuffle paper](https://link.springer.com/content/pdf/10.1007/978-3-319-11212-1_20.pdf) does a nice job of specifying what you really need from such a protocol:

>Second, the
blame phase is entered if the shuffling has not been performed correctly. In that
case, the participants can broadcast their ephemeral decryption keys, along with
the messages they have received. This information allows every participant to
replay the computations of the rest of participants and expose the misbehaving
one. Third, participants could equivocate in the broadcasts of the protocol, e.g.,
by sending different public keys to different participants in the announcement
phase. All participants exchange messages before creating the mixing transaction
to ensure that nobody has equivocated. In case of equivocation, the blame phase
is entered. Since all protocol messages are signed, the equivocating participant
can be identified; two signed messages that are different but belong to the same
sender and the same broadcast step provide evidence of the misbehavior.

I did think briefly about equivocation but I'm not sure if I have any concrete ideas on that yet. I'm not yet convinced you can have a fully 'clean' *blame* protocol using ring sigs instead of the layered encryption design of coinshuffle or the dc-net design of coinshuffle++, but I'm certainly not convinced you can't, either! .. I haven't much of a clue about it to be honest :)

It's also rather telling that nobody every came up with an instantiation of Coinshuffle(++) without a central server (there was a decred project that used a central server for the ++ variant, though I've no idea whether it was really working well or not).

**Comment by supertestnet** [comment]
> So let's ask how the code decides to terminate round 2. Does it just wait for N messages (here 3)? If so, it will terminate on receiving (r3, a3) and move on to round 2, ignoring (r4, a4). Again, since they are anonymous, anyone could send anything at any time, so that certainly seems the most reasonable(?).

No, the code gives everyone 10 seconds to send in their round 2 messages. If the wrong number of messages comes in, it ought to start the blame round to find out which people didn't send in a message or which people sent in too many messages. But I haven't implemented the blame protocol so it doesn't do this yet.

> If so then in round 3 bob will not sign, of course. Is Bob kicked out for refusing to send a transaction-signing message?

If Bob sent a valid message in round 2, he is in no danger of getting kicked out during the transition to round 3. The troll who sent in two cj_addresses will be found out through the blame protocol (not implemented yet) and then everyone will redo round 2 without the troll. Bob "makes it" through the blame protocol without issue because he sent in the right number of messages (and can prove it), then he redoes round 2 correctly because he's not a troll. Then he's in round 3, so he signs the transaction and broadcasts it.

> precisely if we don't use linkable ring sigs, we can't know for sure that they did round 2 correctly

The blame protocol finds them and kicks them out automatically (well it would if I implemented it -- or at least that's what I expect). Either they reveal the private key to their ring key and thus show that they sent in the wrong number of messages, thus "outing" themselves as a troll, or they do not reveal it, thus "outing" themselves as a troll. Once the trolls are identified, everyone just redoes round 2 without them.

> The nice thing about this design for our purposes is that all you have to do to reveal that a ring signature came from your pubkey, is to reveal your private key (so it lines up with your description in the README); the everyone else can calculate your-private-key * other-basepoint to get your key image and verify where it was used.

Sounds cool! I would love to see an implementation of the blame protocol that uses linkable ring signatures instead of the regular kind. Sounds like it could save a round of communication, which is desirable.

**Comment by AdamISZ** [comment]
> > So let's ask how the code decides to terminate round 2. Does it just wait for N messages (here 3)? If so, it will terminate on receiving (r3, a3) and move on to round 2, ignoring (r4, a4). Again, since they are anonymous, anyone could send anything at any time, so that certainly seems the most reasonable(?).
> 
> No, the code gives everyone 10 seconds to send in their round 2 messages. If the wrong number of messages comes in, it ought to start the blame round to find out which people didn't send in a message or which people sent in too many messages. But I haven't implemented the blame protocol so it doesn't do this yet.

Right, I did also think of it that way, i.e. timeout, but, thinking it through more, we still *have* to make linkable ring signatures a requirement for blame to actually work:

Imagine 3 scenarios for what's happened after 10 seconds:

2 messages have arrived - we enter blame phase and should easily be able to kick out the one who didn't send a message (after appropriate further delay). Only problem here is that of course messages on anonymised networks like Tor are often delayed; so in practice it's a bit ugly but it's well within a reasonable design criterion to say "this is our requirement for connectivity/liveness".

3 messages have arrived - but two were from Alice, one from Carol and Bob's was delayed. Here, on entering blame phase, Alice can send a valid proof of validity of *one* of her messages but not the other; again Bob would be kicked out. Again you can not unreasonably argue that Bob has been kicked out with good reason - his connectivity is not good enough, but it's very dodgy that Alice "passes" this round (note that she succeeds in doing so *by omission* of a transcript/proof). Here it's only "dodgy" but :

4 messages have arrived - two were from Alice, one from Carol, one from Bob. In this case in the blame phase Alice sends a proof that message1 is from her and is valid, Bob sends a proof that message3 is from him and is valid, Carol sends a proof that message4 is from her and is valid. Again, Alice's *omission* to reveal that message2 is from her and is "valid" leaves the protocol stuck in apportioning blame.

So linkability is the obvious thing to do to (a) get easy "claimability" and (b) fix the failure mode above, so that it is not possible - every linkable ring signature has to have a key image attached to be valid; if two such r.s. contain the same tag and each individually evaluates as valid, it marks both as "illegal"/invalid (note: you can't forge a valid ring signature with someone else's key and therefore key image, on a new message of your choosing).

**Comment by supertestnet** [comment]
> 4 messages have arrived - two were from Alice, one from Carol, one from Bob. In this case in the blame phase Alice sends a proof that message1 is from her and is valid, Bob sends a proof that message3 is from him and is valid, Carol sends a proof that message4 is from her and is valid. Again, Alice's omission to reveal that message2 is from her and is "valid" leaves the protocol stuck in apportioning blame.

Alice cannot omit proving message2 is from her if she reveals message1 is from her. Revealing herself as the source of message1 requires disclosing her private key from round 1, and that private key reveals that she created the signature she sent in with message2 *and* the signature she sent in with message1. So she is identified as a troll (due to sending two messages in round 2) and all honest parties redo round 2 without her. Bob was also a troll in this round and gets kicked out too, but he didn't mean to, it was due to network delays. Alice and Bob can both try again in ~three bitcoin blocks and if either one is honest this time (including no unreasonably large network delays) they'll make it through just fine.

**Comment by AdamISZ** [comment]
> Alice cannot omit proving message2 is from her if she reveals message1 is from her. Revealing herself as the source of message1 requires disclosing her private key from round 1, and that private key reveals that she created the signature she sent in with message2 and the signature she sent in with message1.

That's exactly where we get into the nitty gritty of "culpability vs exculpability" and "claimability". Simply revealing your private key *does not* prove that you are (or aren't) the author of a particular ring sig in the normal ring signature design (non-linkable; for example, the AOS ring signature or the borromean variant that confidential transactions, mimblewimble etc. used in their range proofs). That's "exculpability", and it's not just a weird side detail, it's considered a very useful property in certain well known applications, like voting schemes. For your use case, as we're discussing here, it's something you *don't* want. That's why I keep bashing on about linkability; with it, you get that opposite case (though it *doesn't* require actually revealing the private key, which is even better but perhaps just a detail).

Edit: and also note there can be other cases, depending on the ring sig algorithm. You might have an algorithm with claimability, i.e. you *can* prove that you're the author of one particular ring sig, without the private key, but still have it not be linkable, because there is no key image which immediately ties two signatures together.

**Comment by supertestnet** [comment]
Thank you for patiently explaining this to me. I am now convinced that linkable ring signatures are the way to go. Thankfully, the library I used also supports linkable ring signatures, so it should be a small code change.
